aten/src/ATen/CUDAGuard.h

```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include <ATen/CUDAStream.h>
+#include <ATen/Context.h>
+#include <ATen/DeviceGuard.h>
+
+#include <cstddef>
+#include <vector>
+
+namespace at {
+
+/// A variant of `DeviceGuard` that augments it with an understanding of CUDA
+/// streams. This guard can not only set and reset the current CUDA device, but
+/// also set and reset the current CUDA stream. It is important to note that
+/// because a CUDA stream is intrinsically associated with the CUDA device to
+/// which it is bound, setting the CUDA stream *also* sets the current CUDA
+/// device to that of the stream.
+struct CUDAGuard {
+  /// Default constructor, does nothing and causes no change in the current
+  /// stream or device until `set_stream` or `set_device` is called.
+  CUDAGuard() = default;
+
+  /// Sets the CUDA stream and its associated device as the current one (calls
+  /// `set_stream`).
+  explicit CUDAGuard(const CUDAStream& stream) {
+    set_stream(stream);
+  }
+
+  /// Calls `set_device` with the given index.
+  explicit CUDAGuard(int32_t device) {
+    set_device(device);
+  }
+
+  /// Copy is disallowed.
+  CUDAGuard(const CUDAGuard&) = delete;
+  CUDAGuard& operator=(const CUDAGuard&) = delete;
+
+  /// Move is disallowed.
+  CUDAGuard(CUDAGuard&&) = delete;
+  CUDAGuard& operator=(CUDAGuard&&) = delete;
+
+  /// Resets the CUDA stream on each device to the one that was active upon
+  /// construction.
+  ~CUDAGuard() {
+    if (!original_streams_.empty()) {
+      for (size_t device = 0; device < original_streams_.size(); ++device) {
+        globalContext().uncheckedSetCurrentCUDAStreamOnDevice(
+            device, original_streams_[device]);
+      }
+    }
+  }
+
+  /// Sets the current CUDA device to the device associated with the given
+  /// stream, and then sets the current stream on that device to the one given.
+  void set_stream(const CUDAStream& stream) {
+    device_guard_.set_index(stream.device());
+    // If we haven't stored the current stream yet, store it now.
+    if (original_streams_.empty()) {
+      const size_t device_count = globalContext().getNumGPUs();
+      original_streams_.reserve(device_count);
+      for (size_t device = 0; device < device_count; ++device) {
+        original_streams_.push_back(
+            globalContext().getCurrentCUDAStreamOnDevice(device));
+      }
+    }
+    globalContext().setCurrentCUDAStreamOnDevice(
+        device_guard_.last_index(), stream);
+  }
+
+  /// Sets the CUDA device to the given one.
+  void set_device(int32_t device) {
+    device_guard_.set_index(device);
+  }
+
+  /// Returns the CUDA stream that was set upon construction of the guard.
```

Comment:
This comment is only accurate if the `CUDAGuard(const CUDAStream& stream)` constructor was used.